### PR TITLE
Resolve NPE merging yaml when resource requests/limits are not set

### PR DESCRIPTION
**Pull request by oliverlockwood**
If you have a `PodTemplate` with a ContainerSpec, and want to also set the `yaml` field to apply some things that aren't directly configurable through this plugin, then it's possible to hit NPEs if resource requests / limits aren't set.  (In Kubernetes, it's perfectly valid not to set resource request / limit values for `cpu` and memory.)

This PR defensively codes to cover that case.

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java [hunk 1]
```diff
@@ -162,6 +149,18 @@ public static Container combine(@CheckForNull Container parent, @Nonnull Contain
         return combined;
     }
 
+    private static Quantity safeGet(Container parent, Container template,
+                                    Function<ResourceRequirements, Map<String, Quantity>> resourceTypeMapper,
+                                    String field) {
+        Quantity templateField = resourceTypeMapper.apply(template.getResources()).get(field);
+        if (templateField != null && !Strings.isNullOrEmpty(templateField.getAmount())) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: would be clearer with `Optional.ofNullable` as in https://github.com/jenkinsci/kubernetes-plugin/pull/311/files#diff-eff844000c9e60765310e70fa8c8a71cR208 ?

Comment (oliverlockwood) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: @carlossg thanks for the feedback.  I've addressed this in the latest commit.

**Comment by therc** [comment]
Is this the proper fix? Either this change or something that was merged later results in objects that match the schema, but are rejected by the API server at validation time. The test case checks the former, but not the latter (which would require the Kubernetes Golang codebase). I suspect the `.orElse(new Quantity(""))` statement.

This is what I am seeing if I don't specify all limits and requests, with line splits added for clarity:
```
Received status: Status(apiVersion=v1, code=400, details=null, kind=Status,
 message=Pod in version "v1" cannot be handled as a Pod: v1.Pod: Spec: v1.PodSpec:
 Containers: []v1.Container: v1.Container: Resources: v1.ResourceRequirements: Limits:
 unmarshalerDecoder: quantities must match the regular expression
 '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$', error found in #10 byte of
 ...|:{"cpu":"","memory":|...,
 bigger context ...|ubectl","ports":[],"resources":{"limits":{"cpu":"","memory":""},"requests":{"cpu":"","memory":""}},"|...,
 metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}),
 reason=BadRequest, status=Failure, additionalProperties={}).
```
The object might be well-formed, but the server rejects it at validation time and returns a 400, at least with Kubernetes 1.10.

Proof that empty strings are problematic:
```
$ kubectl apply -f test.yaml
pod "podtest" created
$ kubectl apply -f test2.yaml
Error from server (BadRequest): error when creating "test2.yaml":
 Pod in version "v1" cannot be handled as a Pod: v1.Pod: Spec: v1.PodSpec:
 Containers: []v1.Container: v1.Container: Resources: v1.ResourceRequirements: Limits:
 unmarshalerDecoder: quantities must match the regular expression
 '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$', error found in #10 byte of
 ...|:{"cpu":""}}}]}}|...,
 bigger context ...|","name":"busybox","resources":{"limits":{"cpu":""}}}]}}|...
$ diff test.yaml test2.yaml
4c4
<   name: podtest
---
>   name: podtest2
10a11,13
>     resources:
>       limits:
>         cpu: ""
```
Reverting to 1.5.2 results in NPEs, of course. 1.6.0 has the same validation errors. I can file an issue in the tracker with all the above.

**Comment by carlossg** [comment]
I think `orElse(null)` could be a better approach if you can test and submit a PR

**Comment by therc** [comment]
I can try, but this is going to be a bit rough, since I haven't written any Java in 15 years. :)

**Comment by therc** [comment]
Ok, #342 seems to work, but it wasn't fun to write using only Chrome as the IDE.

**Pull request merged**